models/block.go

```diff
@@ -93,32 +82,183 @@ func decodeBlockEvent(event cadence.Event) (*Block, error) {
+	Height            uint64
+	Timestamp         uint64
+	TotalSupply       *big.Int
+	ReceiptRoot       gethCommon.Hash
+	TransactionHashes []gethCommon.Hash
+	TotalGasUsed      uint64
 }
 
-func (b *blockV0360) Hash() (gethCommon.Hash, error) {
+func pastBlockHash(b any) (gethCommon.Hash, error) {
 	data, err := rlp.EncodeToBytes(b)
 	return gethCrypto.Keccak256Hash(data), err
 }
+
+// decodeBlockBreakingChanges will try to decode the bytes into all
+// previous versions of block type, if it succeeds it will return the
+// migrated block, otherwise it will return nil.
+func decodeBlockBreakingChanges(encoded []byte) *Block {
+	b0 := &blockV0{}
+	if err := rlp.DecodeBytes(encoded, b0); err == nil {
+		return &Block{
+			Block: &types.Block{
+				ParentBlockHash: b0.ParentBlockHash,
+				Height:          b0.Height,
+				ReceiptRoot:     b0.ReceiptRoot,
+				TotalSupply:     big.NewInt(int64(b0.TotalSupply)),
+			},
+		}
+	}
+
+	b1 := &blockV1{}
+	if err := rlp.DecodeBytes(encoded, b1); err == nil {
+		return &Block{
+			Block: &types.Block{
+				ParentBlockHash: b1.ParentBlockHash,
+				Height:          b1.Height,
+				TotalSupply:     big.NewInt(int64(b1.TotalSupply)),
+				ReceiptRoot:     b1.ReceiptRoot,
+			},
+			TransactionHashes: b1.TransactionHashes,
+		}
+	}
+
+	b2 := &blockV2{}
+	if err := rlp.DecodeBytes(encoded, b2); err == nil {
+		return &Block{
+			Block: &types.Block{
+				ParentBlockHash: b2.ParentBlockHash,
+				Height:          b2.Height,
+				TotalSupply:     big.NewInt(int64(b2.TotalSupply)),
+				ReceiptRoot:     b2.ReceiptRoot,
+			},
+			TransactionHashes: b2.TransactionHashes,
+		}
+	}
+
+	b3 := &blockV3{}
+	if err := rlp.DecodeBytes(encoded, b3); err == nil {
+		return &Block{
+			Block: &types.Block{
+				ParentBlockHash: b3.ParentBlockHash,
+				Height:          b3.Height,
+				TotalSupply:     big.NewInt(int64(b3.TotalSupply)),
+				ReceiptRoot:     b3.ReceiptRoot,
+			},
+			TransactionHashes: b3.TransactionHashes,
+		}
+	}
+
+	b4 := &blockV4{}
+	if err := rlp.DecodeBytes(encoded, b4); err == nil {
+		return &Block{
+			Block: &types.Block{
+				ParentBlockHash: b4.ParentBlockHash,
+				Height:          b4.Height,
+				TotalSupply:     b4.TotalSupply,
+				ReceiptRoot:     b4.ReceiptRoot,
+			},
+			TransactionHashes: b4.TransactionHashes,
+		}
+	}
+
+	b5 := &blockV5{}
+	if err := rlp.DecodeBytes(encoded, b5); err == nil {
+		return &Block{
+			Block: &types.Block{
+				ParentBlockHash: b5.ParentBlockHash,
+				Height:          b5.Height,
+				Timestamp:       b5.Timestamp,
+				TotalSupply:     b5.TotalSupply,
+				ReceiptRoot:     b5.ReceiptRoot,
+			},
+			TransactionHashes: b5.TransactionHashes,
+		}
+	}
+
+	b6 := &blockV6{}
+	if err := rlp.DecodeBytes(encoded, b6); err == nil {
+		return &Block{
+			Block: &types.Block{
+				ParentBlockHash: b5.ParentBlockHash,
+				Height:          b5.Height,
+				Timestamp:       b5.Timestamp,
+				TotalSupply:     b5.TotalSupply,
+				ReceiptRoot:     b5.ReceiptRoot,
```

Comment:
Since:

```bash
// adds total gas used

type blockV6 struct {
```
Shouldn't we populate the `TotalGasUsed` field for `b6` ?

Reply:
yes thank you